### PR TITLE
Site Editor: Support starter patterns

### LINKED ISSUE #55108: Site Editor: creating a new page doesn't invoke pattern selection
When using Twenty Twenty-Four and adding a new page in the Site Editor, despite having patterns registered for page creation, the starter pattern modal never appears causing you to enter an empty page, despite there being patterns registered for this creation process. This stands in contrast to when you create a new page using Pages > Add New Page. 

cc @richtabor for flagging this and @aaronrobertshaw in case you can help :) 

### Visual

https://github.com/WordPress/gutenberg/assets/1813435/0c2738b5-db1b-4f78-92ec-d4d5a33f0144

### PR DESCRIPTION
**Pull request by youknowriad**
Alternative to #55117 
Fixes #55108

## What?

Adds the 'starter patterns' feature to the site editor for users creating a new page.

## Why?
The same feature is available in the post editor when creating a new page, so this provides parity.

## How?

It is slightly trickier in the site editor, as the pattern content needs to be inserted into the 'Content' block when in page editing mode. We can't check for an empty title as well because when creating a page in the site editor we fill the title first. So we're checking for "empty content" to show the modal rather than auto-draft status.

## Testing Instructions

1. Open the site editor
2. Select the 'Pages' option from the navigation
3. Select the 'Draft a new Page' (+ button) option
4. In the resulting modal, give your page a new then select 'Create draft' 
5. The page is created and the 'Choose a Pattern' modal should open
6. Select a pattern, the modal should close and the pattern is inserted into the post content block.
7. Navigate in the site editor to create a new page, the modal should show up again.

Test the modal in the post editor too.

### CODE REVIEW

File: packages/editor/src/components/start-page-options/index.js [hunk 1]
```diff
@@ -3,31 +3,39 @@
...
+			const rootClientId =
+				getRenderingMode() === 'post-only'
+					? ''
+					: getBlocksByName( 'core/post-content' )?.[ 0 ];
 			return {
-				blockPatternsWithPostContentBlockType:
-					getPatternsByBlockTypes( 'core/post-content' ),
+				blockPatternsWithPostContentBlockType: getPatternsByBlockTypes(
+					'core/post-content',
+					rootClientId
```
Comment (youknowriad) [inline packages/editor/src/components/start-page-options/index.js:37]: If we don't pass the right rootClientId, the list of patterns can be empty because we're not allowed to insert patterns top level in "template-locked" mode.

File: packages/editor/src/components/start-page-options/index.js [hunk 2]
```diff
@@ -49,13 +57,21 @@ function useStartPatterns() {
...
+			postId: getCurrentPostId(),
+		};
+	}, [] );
 	return (
 		<BlockPatternsList
 			blockPatterns={ blockPatterns }
 			shownPatterns={ shownBlockPatterns }
 			onClickPattern={ ( _pattern, blocks ) => {
-				resetEditorBlocks( blocks );
+				editEntityRecord( 'postType', postType, postId, { blocks } );
```
Comment (youknowriad) [inline packages/editor/src/components/start-page-options/index.js:74]: This works regardless of the mode, we're actually replacing the patterns of the given post type directly.

**Comment by jasmussen** [comment]
Flow works as intended for me, and matches the admin "add new page" flow. Anything in particular you'd like feedback on? The first time I tried this, there was an odd delay between creating the draft and the modal showing up, but this delay didn't appear the second time so I'm ascribing it to a glitch in the matrix.

**Comment by youknowriad** [comment]
@jasmussen I think the delay is the block patterns being loaded because these load async.

The difference between this and trunk is that now we show this flow for any page that is empty (and not just auto-drafts). The reason is that in the site editor there's no "auto-drafts".

File: packages/block-library/src/block/test/edit.native.js [hunk 3]
```diff
@@ -45,6 +45,7 @@ const getMockedReusableBlock = ( id ) => ( {
 	title: { raw: `Reusable block - ${ id }` },
 	type: 'wp_block',
 	meta: { footnotes: '' },
+	wp_pattern_category: [],
```
Comment (ntsekouras) [inline packages/block-library/src/block/test/edit.native.js:48]: Why this change?

Comment (youknowriad) [reply packages/block-library/src/block/test/edit.native.js:48]: To fix a unit test failure, this field seems mandatory in the REST API

**Comment by ntsekouras** [comment]
By testing in site editor I think it might make sense to add a snackbar when we add patterns. For me the pattern was inserted but in my viewport it wasn't visible due to the template content, so it might confuse folks.

File: packages/editor/src/components/start-page-options/index.js [hunk 4]
```diff
@@ -89,17 +104,29 @@ function StartPageOptionsModal( { onClose } ) {
...
+			isEditedPostEmpty,
+			getCurrentPostType,
+			getCurrentPostId,
+			getEditorSettings,
+		} = select( editorStore );
+		const { __unstableIsPreviewMode: isPreviewMode } = getEditorSettings();
 
-		return (
-			getRenderingMode() === 'post-only' &&
-			! isFeatureActive( 'welcomeGuide' ) &&
```
Comment (ntsekouras) [inline packages/editor/src/components/start-page-options/index.js:98]: We don't want to preserve the welcome guides check?

File: packages/editor/src/components/start-page-options/index.js [hunk 5]
```diff
@@ -89,17 +104,29 @@ function StartPageOptionsModal( { onClose } ) {
...
-		);
+		return {
+			shouldEnableModal:
+				! isPreviewMode && ! isEditedPostDirty() && isEditedPostEmpty(),
+			postType: getCurrentPostType(),
+			postId: getCurrentPostId(),
+		};
 	}, [] );
 
+	useEffect( () => {
```
Comment (ntsekouras) [inline packages/editor/src/components/start-page-options/index.js:125]: Why is this needed?

File: packages/editor/src/components/start-page-options/index.js (hunk 4 above)
Comment (youknowriad) [reply packages/editor/src/components/start-page-options/index.js:98]: It didn't feel that important for me for this PR and it's still and edit-post only preference, so if we want to do that, we'll have to first migrate it in its own PR.

File: packages/editor/src/components/start-page-options/index.js (hunk 5 above)
Comment (youknowriad) [reply packages/editor/src/components/start-page-options/index.js:125]: In the site editor, you can navigate between pages without leaving the page, meaning that if you dismissed the modal for the current page, it shouldn't stay dismissed for other pages.

Comment (ntsekouras) [reply packages/editor/src/components/start-page-options/index.js:125]: Okay. I hadn't tried to dismiss the modal 😄

**Review by ntsekouras — APPROVED** [review]
Looks good, thanks!

**Comment by colorful-tones** [comment]
@youknowriad do you happen to know if this is already noted in a follow up issue, but I just tested this for Gutenberg 18.2 RC1 and it is working fine, but while still in 'draft' it seems the starter pattern modal keeps being triggered when you back out of site editor one step.

Please see this video (near the end around 0:24) for visual of odd behavior. It is not entirely a bug, but I feel like it'll drive folks mad to have the starter pattern modal keep triggering in this flow.

I'm happy to write up an issue if there is not one. Thanks!


https://github.com/WordPress/gutenberg/assets/405912/fd01580c-e5c3-4af5-b9b5-ac5a9e2ba436

**Comment by youknowriad** [comment]
@colorful-tones You're right there's indeed an annoying bug here, I'll follow up with a fix.
